### PR TITLE
Refactor `create_cache`

### PR DESCRIPTION
**Pull request by JoshuaLampert**
This passes the `boundary_conditions` to `create_cache` and dispatches on them instead of looking at the type of the SBP operators to infer whether the problem has periodic or reflecting boundary conditions, e.g. previously it was possible to use `boundary_conditions_reflecting` and pass periodic operators to the `Solver`, which meant that the `create_cache` function also created the cache for the periodic problem instead of for reflecting. This was a bit messy. This way it should work more as expected and is also easier to extend.
On the fly, I put `tmp1`(needed by the `RelaxationCallback`) into the `intitial_cache` instead of always creating it in `create_cache`, which is weird because `tmp1` is sometimes not needed for the discretization itself.
Finally, I fixed the upwind discretization for the reflecting boundary conditions for the `BBMBBMEquations1D`, which was broken before. I also added a test in a similar manner as in #105.

### CODE REVIEW

File: src/equations/bbm_bbm_1d.jl [hunk 1]
```diff
@@ -127,39 +127,43 @@ function source_terms_manufactured_reflecting(q, x, t, equations::BBMBBMEquation
...
     D = equations.D
-    if solver.D1 isa PeriodicDerivativeOperator ||
-       solver.D1 isa UniformPeriodicCoupledOperator ||
-       solver.D1 isa PeriodicUpwindOperators
-        invImD2 = inv(I - 1 / 6 * D^2 * Matrix(solver.D2))
-        return (invImD2 = invImD2, tmp1 = tmp1)
-    elseif solver.D1 isa DerivativeOperator ||
-           solver.D1 isa UniformCoupledOperator ||
-           solver.D1 isa UpwindOperators
+    invImD2 = inv(I - 1 / 6 * D^2 * Matrix(solver.D2))
```
Comment (ranocha) [inline src/equations/bbm_bbm_1d.jl:135]: In the long run, one should consider allowing more flexibility here. For example, FD methods should use sparse matrices and just a factorization instead of the inverse of a dense matrix. But that's not critical right now.

File: src/equations/bbm_bbm_1d.jl [hunk 2]
```diff
@@ -127,39 +127,43 @@ function source_terms_manufactured_reflecting(q, x, t, equations::BBMBBMEquation
...
+function create_cache(mesh, equations::BBMBBMEquations1D,
+                      solver, initial_condition,
+                      ::BoundaryConditionReflecting,
+                      RealT, uEltype)
+    D = equations.D
+    N = nnodes(mesh)
+    M = mass_matrix(solver.D1)
+    Pd = BandedMatrix((-1 => fill(one(real(mesh)), N - 2),), (N, N - 2))
+    D2d = (sparse(solver.D2) * Pd)[2:(end - 1), :]
+    # homogeneous Dirichtlet boundary conditions
```
Comment (ranocha) [inline src/equations/bbm_bbm_1d.jl:148]:
```suggestion
    # homogeneous Dirichlet boundary conditions
```

**Review by ranocha — DISMISSED** [review]
Thanks! I just have some minor comments that can either be resolved quickly or maybe postponed to later development (issues). Feel free to merge :+1:

File: src/equations/bbm_bbm_1d.jl (hunk 1 above)
Comment (JoshuaLampert) [reply src/equations/bbm_bbm_1d.jl:135]: Yeah, you are definitely right. IIRC, when I tried that some time ago, I had some problems making that properly work, but I don't remember the exact issue. Back then I focused on something that just works and is easy to implement andunderstand . But of course this is not the most performant way. I created #107 for that.
